client/src/main/java/io/pravega/client/byteStream/ByteStreamWriter.java

```diff
@@ -0,0 +1,112 @@
+     * 
+     * It is intended that this method not block, but it may in the event that the server becomes
+     * disconnected for sufficiently long or is sufficiently slow that that backlog of data to be
+     * written becomes a memory issue. If this behavior is undesirable the method
+     * {@link #setThrowBeforeBlocking(boolean)} can be used to make this call throw an exception
+     * instead of blocking.
+     * 
+     * @param src The bytes to write.
+     * @throws IOException If for any reason an error occurs writing the data, including if the
+     *             stream is sealed.
+     */
+    public abstract void write(ByteBuffer src) throws IOException;
+
+    /**
+     * Writes the provided data to the segment. The data is buffered internally to avoid blocking.
+     * As such it cannot be assumed to be durably stored until a flush completes.
+     * 
+     * It is intended that this method not block, but it may in the event that the server becomes
+     * disconnected for sufficiently long or is sufficiently slow that that backlog of data to be
+     * written becomes a memory issue. If this behavior is undesirable the method
+     * {@link #setThrowBeforeBlocking(boolean)} can be used to make this call throw an exception
+     * instead of blocking.
+     * 
+     * @see java.io.OutputStream#write(byte[], int, int)
+     */
+    @Override
+    public abstract void write(byte[] b, int off, int len) throws IOException;
+
+    /**
+     * Flushes the buffer and closes the writer. If there is data to flush, this is a blocking
+     * method.
+     * 
+     * @see java.io.OutputStream#close()
+     */
+    @Override
+    public abstract void close() throws IOException;
+
+    /**
+     * Blocks until all data written has been durably persisted.
+     * 
+     * @see java.io.OutputStream#flush()
+     * @throws IOException If for any reason the flush fails including if the stream is sealed.
+     */
+    @Override
+    public abstract void flush() throws IOException;
+
+    /**
+     * Closes the writer similar to {@link #close()} but also seals it so that no future writes can
+     * ever be made.
+     * 
+     * @throws IOException If for any reason the flush fails including if the stream is sealed.
+     */
+    public abstract void closeAndSeal() throws IOException;
+
+    /**
+     * Similar to {@link #flush()} but does not block the calling thread, and instead returns a
+     * future that will be completed when the flush is done.
+     */
+    public abstract CompletableFuture<Void> flushAsync();
+
+    /**
+     * This makes a synchronous RPC call to the server to obtain the total number of bytes written
+     * to the segment in its history. This is the sum total of the bytes written in all calls to
+     * {@link #write(byte[])} that have been flushed. It does not include data that was passed to
+     * {@link #write(byte[])} but which has not yet been persisted.
+     */
+    public abstract long fetchOffset();
```

Comment:
I don't like the term 'length' as it implies that it might decrease if the data get's truncated. 
I don't like 'writeOffset' in this case because it implies that this is the offset which new data will be written if you call 'write' this is not correct, because there may be data still buffered by this class not included. 
I originally called it 'fetchPersistedOffset' in an attempt to clarify this. However @fpj did not like this. 
I am happy to add another word in there, but I honestly don't have a good one. Suggestions?

Reply:
Is this equivalent to `fetchTailOffset` in the reader? Should we consider calling the same here for symmetry?

Reply:
yes. It is the same. I can call it that if that is preferred.